lib/gems/pending/util/miq_file_storage.rb

```diff
@@ -0,0 +1,203 @@
+      reset_vars
+    end
+    alias upload add
+
+    def mkdir(dir) # rubocop:disable Lint/UnusedMethodArgument
+      raise NotImplementedError, "#{self.class}##{__callee__} is not defined"
+    end
+
+    # :call-seq:
+    #   download( local_io, remote_uri )
+    #   download( nil,      remote_uri ) { |input_writer| ... }
+    #
+    # Download a file from a remote uri.
+    #
+    # In non-block form, the remote_uri is saved to the local_io.
+    #
+    # In block form, the local_io is omitted, and it is set to a PTY writer
+    # path that will assumed to be read by the block provided.
+    def download(local_file, remote_file_uri, &block)
+      @remote_file_path = remote_file_uri
+      if block_given?
+        thread = handle_io_block(&block)
+        download_single(remote_file_uri, input_writer)
+        input_writer.close
+        thread.join
+      else
+        download_single(remote_file_uri, local_file)
+      end
+    ensure
+      reset_vars
+    end
+
+    private
+
+    def upload_single(file) # rubocop:disable Lint/UnusedMethodArgument
+      raise NotImplementedError, "#{self.class}#upload_single is not defined"
+    end
+
+    def upload_splits
+      @position = 0
+      until source_input.eof?
+        upload_single(next_split_filename)
+        @position += byte_count
+      end
+    end
+
+    def initialize_upload_vars(*upload_args)
+      upload_args.pop if (@byte_count = parse_byte_value(upload_args.last))
+      @remote_file_path = upload_args.pop
+
+      unless upload_args.empty?
+        source        = upload_args.pop
+        @source_input = source.kind_of?(IO) ? source : File.open(source, "r")
+      end
+    end
+
+    def parse_byte_value(bytes)
+      match = bytes.to_s.match(BYTE_HASH_MATCH) || return
+
+      bytes = match[:BYTE_NUM].to_f
+      if match[:BYTE_QUALIFIER]
+        bytes *= BYTE_HASH[match[:BYTE_QUALIFIER].downcase]
+      end
+      bytes.to_i
+    end
+
+    def handle_io_block
+      if block_given?
+        require "tmpdir"
+
+        # create pathname, but don't create the file for it (next line)
+        fifo_path = Pathname.new(Dir::Tmpname.create("") {})
+        File.mkfifo(fifo_path)
+
+        # For #Reasons(TM), the reader must be opened first
+        @source_input = File.open(fifo_path.to_s, IO::RDONLY | IO::NONBLOCK)
+        @input_writer = File.open(fifo_path.to_s, IO::WRONLY | IO::NONBLOCK)
+
+        Thread.new do
+          yield fifo_path      # send the path to the block to get executed
+          @input_writer.close  # close the file so we know we hit EOF (for #add)
+        end
+      end
+    end
+
+    def reset_vars
+      File.delete(@input_writer.path) if @input_writer
+      @position, @byte_count, @remote_file_path, @source_input, @input_writer = nil
+    end
+
+    def next_split_filename
+      "#{remote_file_path}.#{'%05d' % (@position / byte_count + 1)}"
+    end
+
+    def download_single(source, destination) # rubocop:disable Lint/UnusedMethodArgument
```

Comment:
How does the download side handle splitting? Is there some kind of detection that the file being requested is split on the remote side?

Reply:
Yeah, I was thinking about this, but I decided to not implement this for the first pass.

The short answer is file splitting was originally targeted for support when uploading to a endpoint that had an upload size limit (might want to have @dmetzger57 weigh in, though, as this was technically his "feature creep suggestion").  Using the split versions just happened to be something that worked for other storage backends without issue, so hooking them up was a zero cost.

I _**COULD**_ try and do auto detection, but I think that might be a bit of a stretch to squeeze this in on time.  Leaning more towards skipping that for now and seeing if this is something we can add later.  I will at the very least research what is needed to handle this.

Thoughts on that path going forward?

Reply:
My only concern with this is if we do file splitting for database backups, not having this could make restore very difficult to implement.

So if the only file store that will have splitting by default for now is going to be FTP then this should be fine, but if we're doing splitting for other storage types we need to solve this so that we can restore a split backup.

Reply:
> So if the only file store that will have splitting by default for now is going to be FTP then this should be fine...

@carbonin So everything is going to have splitting, it is baked into the the question set on `DatabaseAdmin` in `appliance_console` and the option params in the rake tasks, aside from the fact that it is part of the top level logic in `MiqFileStorage`, so I am not planning on doing any logic to make it available for one provider only.

So this said, I don't see this as being "easy" to implement auto detection but agree that this is worthwhile to have.  To me this sees like something I could follow up and do in a subsequent PR.  Does that sound reasonable or is this something that needs to be included for this PR to be considered merge-able?

cc @dmetzger57

Reply:
My request for this PR is to not introduce restore-side regressions for existing backup targets. If FTP can't restore backups, that's fine, but if this breaks restore from NFS, that's not fine.

So, hypothetically, if we do a backup to an NFS target, how does restore work when the backup is split on the remote? I may be missing some crucial detail, but it feels like this won't work as is.

Reply:
@carbonin Sorry, I think I now see where the confusion is.  So to start off:

> My request for this PR is to not introduce restore-side regressions for existing backup targets.

There are none, but I think I misunderstood where your confusion was possibly originating from.

`MiqFileStorage#download_single` is used in both cases of `#download` (block form and not).  This differs from `#add` which will call either `#upload_splits` when a `byte_count` is passed, calling `#upload_single` multiple times, or just directly to `#upload_single`.

`#download` should work as it has previously.

> So, hypothetically, if we do a backup to an NFS target, how does restore work when the backup is split on the remote?

This, however won't work, but again, splitting is "opt-in", and isn't something you are required to do.  That means:

- You can still upload a single file backup and restore that single file backup as you had in the past.
- You can also upload files split on file size
- Both methods will be streamed through the same code

What you can't do is restore from a remote that is split automatically right now, and I am not even sure how that would look if we did (do we use a glob?  Do we try and `ls` for the file if an exact match isn't found?  Do we have to re-prompt the user if we find split files?).

You would have to cat the files together on said remote and then restore currently.

* * *

I see being able to upload a file split as a new feature, and being able to restore split files as an augment to said feature, or a new feature in it of itself.  The original feature was in place so Support could upload a part based dump automatically.  You can always skip splitting the file (by saying 'n'), as shown on how it is prompted in the integrating PR in the [`appliance_console`](https://github.com/ManageIQ/manageiq-appliance_console/pull/50):

```
Would you like to split the dump output into multiple parts? (Y/N): y
Enter the byte size to split by: |500M| 750M
```

Possibly a warning of "currently there is no support automatically restore split backups/dumps" might be a good idea, but again, I don't see that as a "regression", just missing functionality.

Does that make sense?

Reply:
Ah okay, I was missing the distinction between "everything is going to have splitting" and "splitting is opt-in".

Like I said, as long as this doesn't change the behavior for existing targets it's fine with me. :+1: